File: docs/examples/jdbc-agent-template/src/main/java/com/dbx/agent/template/TemplateAgent.java
package com.dbx.agent.template;

import com.dbx.agent.BaseDatabaseAgent;
import com.dbx.agent.ColumnInfo;
import com.dbx.agent.ConnectParams;
import com.dbx.agent.DatabaseInfo;
import com.dbx.agent.ExecuteQueryOptions;
import com.dbx.agent.ForeignKeyInfo;
import com.dbx.agent.IndexInfo;
import com.dbx.agent.JdbcExecutor;
import com.dbx.agent.JdbcIdentifiers;
import com.dbx.agent.JsonRpcServer;
import com.dbx.agent.QueryResult;
import com.dbx.agent.TableInfo;
import com.dbx.agent.TriggerInfo;
import java.sql.Connection;
import java.sql.DatabaseMetaData;
import java.sql.DriverManager;
import java.sql.ResultSet;
import java.util.ArrayList;
import java.util.Collections;
import java.util.Comparator;
import java.util.LinkedHashMap;
import java.util.LinkedHashSet;
import java.util.List;
import java.util.Map;
import java.util.Set;

public final class TemplateAgent extends BaseDatabaseAgent {
    private static final String DRIVER_CLASS = "com.example.jdbc.TemplateDriver";
    private Connection connection;
    private String databaseName = "";

    @Override
    public Connection getConnection() {
        return connection;
    }

    @Override
    public void connect(ConnectParams params) {
        uncheckedVoid(() -> {
            Class.forName(DRIVER_CLASS);
            connection = DriverManager.getConnection(jdbcUrl(params), params.getUsername(), params.getPassword());
            databaseName = params.getDatabase();
        });
    }

    @Override
    public boolean testConnection(ConnectParams params) {
        return unchecked(() -> {
            Class.forName(DRIVER_CLASS);
            try (Connection conn = DriverManager.getConnection(jdbcUrl(params), params.getUsername(), params.getPassword())) {
                return conn.isValid(5);
            }
        });
    }

    @Override
    public List<DatabaseInfo> listDatabases() {
        return Collections.singletonList(new DatabaseInfo(databaseName.isBlank() ? "default" : databaseName));
    }

    @Override
    public List<String> listSchemas() {
        return unchecked(() -> {
            List<String> result = new ArrayList<>();
            try (ResultSet rs = requireConnected().getMetaData().getSchemas()) {
                while (rs.next()) {
                    result.add(rs.getString("TABLE_SCHEM"));
                }
            }
            Collections.sort(result);
            return result;
        });
    }

    @Override
    public List<TableInfo> listTables(String schema) {
        return unchecked(() -> {
            List<TableInfo> result = new ArrayList<>();
            try (ResultSet rs = requireConnected().getMetaData().getTables(null, schema, "%", new String[]{"TABLE", "VIEW"})) {
                while (rs.next()) {
                    result.add(new TableInfo(rs.getString("TABLE_NAME"), rs.getString("TABLE_TYPE"), null));
                }
            }
            result.sort(Comparator.comparing(TableInfo::getName));
            return result;
        });
    }

    @Override
    public List<ColumnInfo> getColumns(String schema, String table) {
        return unchecked(() -> {
            DatabaseMetaData meta = requireConnected().getMetaData();
            Set<String> primaryKeys = new LinkedHashSet<>();
            try (ResultSet rs = meta.getPrimaryKeys(null, schema, table)) {
                while (rs.next()) {
                    primaryKeys.add(rs.getString("COLUMN_NAME"));
                }
            }

            List<ColumnInfo> result = new ArrayList<>();
            try (ResultSet rs = meta.getColumns(null, schema, table, "%")) {
                while (rs.next()) {
                    String name = rs.getString("COLUMN_NAME");
                    Integer size = intOrNull(rs, "COLUMN_SIZE");
                    result.add(new ColumnInfo(
                        name,
                        rs.getString("TYPE_NAME"),
                        rs.getInt("NULLABLE") == DatabaseMetaData.columnNullable,
                        rs.getString("COLUMN_DEF"),
                        primaryKeys.contains(name),
                        null,
                        rs.getString("REMARKS"),
                        size,
                        intOrNull(rs, "DECIMAL_DIGITS"),
                        size
                    ));
                }
            }
            return result;
        });
    }

    @Override
    public List<IndexInfo> listIndexes(String schema, String table) {
        return unchecked(() -> {
            Map<String, List<String>> columnsByIndex = new LinkedHashMap<>();
            Map<String, Boolean> uniqueByIndex = new LinkedHashMap<>();

            try (ResultSet rs = requireConnected().getMetaData().getIndexInfo(null, schema, table, false, false)) {
                while (rs.next()) {
                    String name = rs.getString("INDEX_NAME");
                    String column = rs.getString("COLUMN_NAME");
                    if (name == null || column == null) {
                        continue;
                    }
                    columnsByIndex.computeIfAbsent(name, ignored -> new ArrayList<>()).add(column);
                    uniqueByIndex.put(name, !rs.getBoolean("NON_UNIQUE"));
                }
            }

            List<IndexInfo> result = new ArrayList<>();
            for (Map.Entry<String, List<String>> entry : columnsByIndex.entrySet()) {
                String name = entry.getKey();
                result.add(new IndexInfo(
                    name,
                    entry.getValue(),
                    uniqueByIndex.getOrDefault(name, false),
                    false,
                    null,
                    null,
                    null,
                    null
                ));
            }
            return result;
        });
    }

    @Override
    public List<ForeignKeyInfo> listForeignKeys(String schema, String table) {
        return unchecked(() -> {
            List<ForeignKeyInfo> result = new ArrayList<>();
            try (ResultSet rs = requireConnected().getMetaData().getImportedKeys(null, schema, table)) {
                while (rs.next()) {
                    result.add(new ForeignKeyInfo(
                        rs.getString("FK_NAME"),
                        rs.getString("FKCOLUMN_NAME"),
                        rs.getString("PKTABLE_NAME"),
                        rs.getString("PKCOLUMN_NAME")
                    ));
                }
            }
            return result;
        });
    }

    @Override
    public List<TriggerInfo> listTriggers(String schema, String table) {
        return Collections.emptyList();
    }

    @Override
    public QueryResult executeQuery(String sql, String schema, ExecuteQueryOptions options) {
        return JdbcExecutor.INSTANCE.execute(
            requireConnected(),
            sql,
            schema,
            this::setSchemaSQL,
            options.getMaxRows(),
            options.getFetchSize(),
            JdbcExecutor.INSTANCE::defaultResultValue
        );
    }

    @Override
    public String setSchemaSQL(String schema) {
        return "SET SCHEMA " + JdbcIdentifiers.INSTANCE.doubleQuote(schema);
    }

    @Override
    public void disconnect() {
        uncheckedVoid(() -> {
            if (connection != null) {
                connection.close();
            }
            connection = null;
        });
    }

    private static String jdbcUrl(ConnectParams params) {
        return "jdbc:template://" + params.getHost() + ":" + params.getPort() + "/" + params.getDatabase();
    }

    private static Integer intOrNull(ResultSet rs, String column) throws Exception {
        Object value = rs.getObject(column);
        return value instanceof Number ? ((Number) value).intValue() : null;
    }

    public static void main(String[] args) {
        new JsonRpcServer(new TemplateAgent()).run();
    }
}

File: docs/examples/jdbc-agent-template/src/main/kotlin/com/dbx/agent/template/TemplateAgent.kt
package com.dbx.agent.template

import com.dbx.agent.ColumnInfo
import com.dbx.agent.ConnectParams
import com.dbx.agent.DatabaseAgent
import com.dbx.agent.DatabaseInfo
import com.dbx.agent.ForeignKeyInfo
import com.dbx.agent.IndexInfo
import com.dbx.agent.JdbcExecutor
import com.dbx.agent.JdbcIdentifiers
import com.dbx.agent.JsonRpcServer
import com.dbx.agent.QueryResult
import com.dbx.agent.TableInfo
import com.dbx.agent.TriggerInfo
import java.sql.Connection
import java.sql.DriverManager
import java.sql.ResultSet

class TemplateAgent : DatabaseAgent {
    private var connection: Connection? = null
    private var databaseName: String = ""

    override fun getConnection(): Connection? = connection

    override fun connect(params: ConnectParams) {
        Class.forName(DRIVER_CLASS)
        connection = DriverManager.getConnection(jdbcUrl(params), params.username, params.password)
        databaseName = params.database
    }

    override fun testConnection(params: ConnectParams): Boolean {
        Class.forName(DRIVER_CLASS)
        DriverManager.getConnection(jdbcUrl(params), params.username, params.password).use { conn ->
            return conn.isValid(5)
        }
    }

    override fun listDatabases(): List<DatabaseInfo> {
        return listOf(DatabaseInfo(databaseName.ifBlank { "default" }))
    }

    override fun listSchemas(): List<String> {
        val conn = requireConnection()
        return conn.metaData.schemas.use { rs ->
            buildList {
                while (rs.next()) {
                    add(rs.getString("TABLE_SCHEM"))
                }
            }.sorted()
        }
    }

    override fun listTables(schema: String): List<TableInfo> {
        val conn = requireConnection()
        val types = arrayOf("TABLE", "VIEW")
        return conn.metaData.getTables(null, schema, "%", types).use { rs ->
            buildList {
                while (rs.next()) {
                    add(TableInfo(rs.getString("TABLE_NAME"), rs.getString("TABLE_TYPE")))
                }
            }.sortedBy { it.name }
        }
    }

    override fun getColumns(schema: String, table: String): List<ColumnInfo> {
        val conn = requireConnection()
        val primaryKeys = conn.metaData.getPrimaryKeys(null, schema, table).use { rs ->
            buildSet {
                while (rs.next()) {
                    add(rs.getString("COLUMN_NAME"))
                }
            }
        }

        return conn.metaData.getColumns(null, schema, table, "%").use { rs ->
            buildList {
                while (rs.next()) {
                    val name = rs.getString("COLUMN_NAME")
                    add(
                        ColumnInfo(
                            name = name,
                            data_type = rs.getString("TYPE_NAME"),
                            is_nullable = rs.getInt("NULLABLE") == java.sql.DatabaseMetaData.columnNullable,
                            column_default = rs.getString("COLUMN_DEF"),
                            is_primary_key = name in primaryKeys,
                            extra = null,
                            comment = rs.getString("REMARKS"),
                            numeric_precision = rs.getObject("COLUMN_SIZE")?.let { (it as Number).toInt() },
                            numeric_scale = rs.getObject("DECIMAL_DIGITS")?.let { (it as Number).toInt() },
                            character_maximum_length = rs.getObject("COLUMN_SIZE")?.let { (it as Number).toInt() },
                        )
                    )
                }
            }
        }
    }

    override fun listIndexes(schema: String, table: String): List<IndexInfo> {
        val conn = requireConnection()
        val columnsByIndex = linkedMapOf<String, MutableList<String>>()
        val uniqueByIndex = mutableMapOf<String, Boolean>()

        conn.metaData.getIndexInfo(null, schema, table, false, false).use { rs ->
            while (rs.next()) {
                val name = rs.getString("INDEX_NAME") ?: continue
                val column = rs.getString("COLUMN_NAME") ?: continue
                columnsByIndex.getOrPut(name) { mutableListOf() }.add(column)
                uniqueByIndex[name] = !rs.getBoolean("NON_UNIQUE")
            }
        }

        return columnsByIndex.map { (name, columns) ->
            IndexInfo(
                name = name,
                columns = columns,
                is_unique = uniqueByIndex[name] ?: false,
                is_primary = false,
                index_type = null,
            )
        }
    }

    override fun listForeignKeys(schema: String, table: String): List<ForeignKeyInfo> {
        val conn = requireConnection()
        return conn.metaData.getImportedKeys(null, schema, table).use { rs ->
            buildList {
                while (rs.next()) {
                    add(
                        ForeignKeyInfo(
                            name = rs.getString("FK_NAME"),
                            column = rs.getString("FKCOLUMN_NAME"),
                            ref_table = rs.getString("PKTABLE_NAME"),
                            ref_column = rs.getString("PKCOLUMN_NAME"),
                        )
                    )
                }
            }
        }
    }

    override fun listTriggers(schema: String, table: String): List<TriggerInfo> {
        return emptyList()
    }

    override fun executeQuery(sql: String, schema: String?): QueryResult {
        return JdbcExecutor.execute(requireConnection(), sql, schema, ::setSchemaSQL)
    }

    override fun setSchemaSQL(schema: String): String {
        return "SET SCHEMA ${JdbcIdentifiers.doubleQuote(schema)}"
    }

    override fun disconnect() {
        connection?.close()
        connection = null
    }

    private fun requireConnection(): Connection {
        return connection ?: throw IllegalStateException("Not connected")
    }

    private fun jdbcUrl(params: ConnectParams): String {
        return "jdbc:template://${params.host}:${params.port}/${params.database}"
    }

    private fun stringResultValue(rs: ResultSet, index: Int, sqlType: Int): Any? {
        return rs.getString(index)
    }

    private companion object {
        const val DRIVER_CLASS = "com.example.jdbc.TemplateDriver"
    }
}

fun main() {
    JsonRpcServer(TemplateAgent()).run()
}
